src/util/lineItems.js

```diff
@@ -0,0 +1,168 @@
+  const numericPrice = convertMoneyToNumber(unitPrice);
+  const numericTotalPrice = new Decimal(numericPrice).times(unitCount).toNumber();
+  return new Money(
+    convertUnitToSubUnit(numericTotalPrice, unitDivisor(unitPrice.currency)),
+    unitPrice.currency
+  );
+};
+
+const calculateTotalPriceFromPercentage = (unitPrice, percentage) => {
+  const numericPrice = convertMoneyToNumber(unitPrice);
+  const numericTotalPrice = new Decimal(numericPrice)
+    .times(percentage)
+    .dividedBy(100)
+    .toNumber();
+  return new Money(
+    convertUnitToSubUnit(numericTotalPrice, unitDivisor(unitPrice.currency)),
+    unitPrice.currency
+  );
+};
+
+const calculateTotalPriceFromSeats = (unitPrice, unitCount, seats) => {
+  const numericPrice = convertMoneyToNumber(unitPrice);
+  const numericTotalPrice = new Decimal(numericPrice)
+    .times(unitCount)
+    .times(seats)
+    .toNumber();
+  return new Money(
+    convertUnitToSubUnit(numericTotalPrice, unitDivisor(unitPrice.currency)),
+    unitPrice.currency
+  );
+};
+
+export const validLineItem = params => {
+  const { code, unitPrice, quantity, percentage, seats, units, includeFor } = params;
+
+  let pricingParams;
+
+  if (quantity) {
+    pricingParams = { quantity, lineTotal: calculateTotalPriceFromQuantity(unitPrice, quantity) };
+  } else if (percentage) {
+    pricingParams = {
+      percentage,
+      lineTotal: calculateTotalPriceFromPercentage(unitPrice, percentage),
+    };
+  } else if (seats && units) {
+    pricingParams = {
+      seats,
+      units,
+      lineTotal: calculateTotalPriceFromSeats(unitPrice, units, seats),
+    };
+  } else {
+    console.error(
+      "Can't calculate the lineTotal of lineItem: ",
+      code,
+      ' Make sure you have provided quantity, percentage or both seats and units'
+    );
+  }
+
+  return {
+    code: code,
+    unitPrice,
+    includeFor: includeFor ? includeFor : ['customer', 'provider'],
+    ...pricingParams,
+  };
+};
+
+export const calculateTotalFromLineItems = lineItems => {
+  const numericTotalPrice = lineItems.reduce((sum, lineItem) => {
+    const numericPrice = convertMoneyToNumber(lineItem.lineTotal);
+    return new Decimal(numericPrice).add(sum);
+  }, 0);
+
+  const unitPrice = lineItems[0].unitPrice;
+
+  return new Money(
+    convertUnitToSubUnit(numericTotalPrice.toNumber(), unitDivisor(unitPrice.currency)),
+    unitPrice.currency
+  );
+};
+
+export const calculateTotalForProvider = lineItems => {
+  const providerLineItems = lineItems.filter(lineItem => lineItem.includeFor.includes('provider'));
+  return calculateTotalFromLineItems(providerLineItems);
+};
+
+export const calculateTotalForCustomer = lineItems => {
+  const providerLineItems = lineItems.filter(lineItem => lineItem.includeFor.includes('customer'));
+  return calculateTotalFromLineItems(providerLineItems);
+};
+
+export const constructLineItems = (startDate, endDate, unitPrice) => {
+  const bookingLineItem = validLineItem({
```

Comment:
Instead of wrapping each line item with the `validLineItem` function, you could collect all line items into an array, and then iterate each item with the function:

```js
const lineItem1 = { /* ... */ };
const lineItem2 = { /* ... */ };
const lineItem3 = { /* ... */ };

const lineItems = [
  lineItem1,
  lineItem2,
  lineItem3,
];

return lineItems.map(validLineItem);
```

You can even skip adding each item into a variable, and just construct them inline within the `lineItems` array.

Reply:
The reason why they are now wrapped separately is that some lineItems (e.g. commissions) need to know the lineTotal of other lineItems and that's calculated inside validLineItem function. 

Maybe a better solution would be to add new helper function for calculating the lineTotal and have other validations separately for all of the lineItems like in your example.

Reply:
Ah ok. Validation separate from line item creation would be nice, but also not a huge problem in the current solution.